### ISSUE TITLE
How to properly set email polling

### ISSUE THREAD

**Issue by floatingpurr** [django-helpdesk/django-helpdesk#590]
Hello there, I am wondering which is the correct way to set up the email polling for updating ticket follow-ups. Reading the docs, it seems like a cron job is needed. Nevertheless, in the admin form for queues there are some settings related to this aim (e.g., Allow E-Mail Submission? E-Mail Check Interval, etc...).

Do I need to set up a cron job or can I just work with the queue settings?

**Update 1**
I tried to set up queue settings using IMAP. Since I want to poll just for update I set also `QUEUE_EMAIL_BOX_UPDATE_ONLY = True `. Now, when I run `manage.py get_email` for testing, a full scan of the Inbox starts:

```
Attempting IMAP server login
Received 285 messages from IMAP server
Processing message 1
No tracking ID matched.
Message 1 was not successfully processed, and will be left on IMAP server
Processing message 2
No tracking ID matched.
Message 2 was not successfully processed, and will be left on IMAP server
Processing message 3
etc...
```

But later it fails. Here is the traceback:

```
No tracking ID matched.
Traceback (most recent call last):
  File "./manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 77, in handle
    process_email(quiet=quiet)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 110, in process_email
    process_queue(q, logger=logger)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 232, in process_queue
    ticket = ticket_from_message(message=full_message, queue=q, logger=logger)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 390, in ticket_from_message
    mail = BeautifulSoup(part.get_payload(), "lxml")
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/__init__.py", line 225, in __init__
    markup, from_encoding, exclude_encodings=exclude_encodings)):
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/builder/_lxml.py", line 118, in prepare_markup
    for encoding in detector.encodings:
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/dammit.py", line 257, in encodings
    self.markup, self.is_html)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/dammit.py", line 315, in find_declared_encoding
    declared_encoding_match = xml_encoding_re.search(markup, endpos=xml_endpos)
TypeError: expected string or buffer
```
Log file:
```
Processing message 35
No tracking ID matched.
Discovered plain text MIME part
Discovered HTML MIME part
Try to base64 decode the attachment payload
Found MIME attachment part-2None
Discovered plain text MIME part
Discovered plain text MIME part
Try to base64 decode the attachment payload
Found MIME attachment part-5.mht
```
Python 2.7. Libraries version:
```
beautifulsoup4==4.6.0
Django==1.11.9
django-bootstrap-form==3.3
django-helpdesk==0.2.6
django-markdown-deux==1.0.5
email-reply-parser==0.5.9
lxml==4.1.1
markdown2==2.3.5
MySQL-python==1.2.5
pytz==2017.3
simplejson==3.13.2
six==1.11.0
wheel==0.26.0
````

**Update 2**

It seems the problem lies in `get_email.py` in the following lines:
```Python
    if not body:
        mail = BeautifulSoup(part.get_payload(), "lxml")
        if ">" in mail.text:
            message_body = mail.text.split(">")[1]
            body = message_body.encode('ascii', errors='ignore')
        else:
            body = mail.text
```

In the reported exception, `part.get_payload()`returns en empty list. It happens for some postmaster automatic replies (e.g., automatic replies in case of undeliverable emails). Probably it happens since in those cases the last element in `part.get_payload()` has been popped here:

```Python
    if isinstance(payload, list):
        payload = payload.pop().as_string()
```

I don't understand this behavior.

I could ignore "toxic" mail senders but this is just a temporary and unsafe workaround. How can I permantely fix this problem?

Bye!

**Comment by floatingpurr** [comment]
Follow up: I tried to set up queue settings using IMAP. Since I want to poll just for update I set also `QUEUE_EMAIL_BOX_UPDATE_ONLY = True `. Now, when I run `manage.py get_email` for testing a full scan of the Inbox starts:

```
Attempting IMAP server login
Received 285 messages from IMAP server
Processing message 1
No tracking ID matched.
Message 1 was not successfully processed, and will be left on IMAP server
Processing message 2
No tracking ID matched.
Message 2 was not successfully processed, and will be left on IMAP server
Processing message 3
etc...
```

But later it fails. Here is the traceback:

```
No tracking ID matched.
Traceback (most recent call last):
  File "./manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 77, in handle
    process_email(quiet=quiet)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 110, in process_email
    process_queue(q, logger=logger)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 232, in process_queue
    ticket = ticket_from_message(message=full_message, queue=q, logger=logger)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/helpdesk/management/commands/get_email.py", line 390, in ticket_from_message
    mail = BeautifulSoup(part.get_payload(), "lxml")
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/__init__.py", line 225, in __init__
    markup, from_encoding, exclude_encodings=exclude_encodings)):
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/builder/_lxml.py", line 118, in prepare_markup
    for encoding in detector.encodings:
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/dammit.py", line 257, in encodings
    self.markup, self.is_html)
  File "/foo/bar/.myVirtualenvsPath/myVenv/local/lib/python2.7/site-packages/bs4/dammit.py", line 315, in find_declared_encoding
    declared_encoding_match = xml_encoding_re.search(markup, endpos=xml_endpos)
TypeError: expected string or buffer
```

How can I fix this problem?

Bye!

**Comment by gwasser** [comment]
Our `get_email` script has undergone revision to update it for python 3, so we're working out the bugs still. It almost certainly needs better exception handling. Thanks for pointing this out!

**Comment by nigel2357** [comment]
Here's a fix for this problem, which I also ran into. At the moment it's of the nature of a hotfix. It simply ignores the error and carries on processing the other e-mails, so the mail message(s) that would otherwise crash the cron job simply accumulate in the inbox until somebody deals with them manually.

A better fix, which I might write later, would forward them to a designated manager address.

I'm new to git and contributing to a project like this. Is this patch an appropriate way of offering this fix?

diff --git a/helpdesk/management/commands/get_email.py b/helpdesk/management/commands/get_email.py
index f2e1108..8ba8e55 100755
--- a/helpdesk/management/commands/get_email.py
+++ b/helpdesk/management/commands/get_email.py
@@ -229,7 +229,10 @@ def process_queue(q, logger):
                 logger.info("Processing message %s" % num)
                 status, data = server.fetch(num, '(RFC822)')
                 full_message = encoding.force_text(data[0][1], errors='replace')
-                ticket = ticket_from_message(message=full_message, queue=q, logger=logger)
+                try:
+                    ticket = ticket_from_message(message=full_message, queue=q, logger=logger)
+                except TypeError:
+                    ticket = None  # hotfix. Need to work out WHY.
                 if ticket:
                     server.store(num, '+FLAGS', '\\Deleted')
                     logger.info("Successfully processed message %s, deleted from IMAP server" % num)

**Comment by gwasser** [comment]
Thank you for the hotfix! Better than crashing, but agreed we need a better long term solution. I applied the patch to `master`. If you prefer diffs for small things that's ok, but probably better to send a pull request instead. GitHub has some info on how to do that, hopefully it helps but let me know if you have questions: https://help.github.com/articles/creating-a-pull-request/

**Comment by spamandeggs** [comment]
Hi,
here is a quick fix for a bug I got when using manage.py get_email : 
I'm using python 3.4.7 django 1.11 helpdesk 0.2.6
the list server.retr(msgNum)[1] that encoding.force_text() tries to join contains bytes, not strings, so the console throws an encoding.force_text traceback.

below is what I did, only for pop3, _not imap_, but it looks fine now in my project.
I suppose python 2.7 will be ok with this mod
(but I still get weird encoding for html attachement :s)

```
in helpdesk/management/commands/get_email.py,  process_queue() :

            msgNum = msg.split(" ")[0]
            logger.info("Processing message %s" % msgNum)
            # added this check :
            raw_content = server.retr(msgNum)[1]
            if type(raw_content[0]) == bytes:
                full_message =  "\n".join([elm.decode('utf-8') for elm in raw_content])
            else :
                full_message = encoding.force_text("\n".join(raw_content), errors='replace')
            #
            ticket = ticket_from_message(message=full_message, queue=q, logger=logger)
```

**Comment by gwasser** [comment]
The huge differences in string and encoding handling between python 2 and 3 are the problems here. The code base was originally for 2 and I've been trying to whip it into shape for py3. I'll probably use `six` to add your changes for python 3 and keep the old code for python 2 (as it likely does still work fine on py2). Thanks!

**Comment by gwasser** [comment]
Your change was added @spamandeggs but I forgot to update the thread apparently, oops! Let me know if `master` works for you but I think this is fixed for now (or at least patched, the 0.3 release will require Django 2 and therefore Python 3 and free us from the need to use `six` and deal with string differences).

**Comment by timthelion** [comment]
This is almost certainly fixed
